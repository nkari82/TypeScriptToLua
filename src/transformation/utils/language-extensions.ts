import * as ts from "typescript";
import { TransformationContext } from "../context";

export enum ExtensionKind {
    MultiFunction = "MultiFunction",
    RangeFunction = "RangeFunction",
    VarargConstant = "VarargConstant",
    AdditionOperatorType = "Addition",
    AdditionOperatorMethodType = "AdditionMethod",
    SubtractionOperatorType = "Subtraction",
    SubtractionOperatorMethodType = "SubtractionMethod",
    MultiplicationOperatorType = "Multiplication",
    MultiplicationOperatorMethodType = "MultiplicationMethod",
    DivisionOperatorType = "Division",
    DivisionOperatorMethodType = "DivisionMethod",
    ModuloOperatorType = "Modulo",
    ModuloOperatorMethodType = "ModuloMethod",
    PowerOperatorType = "Power",
    PowerOperatorMethodType = "PowerMethod",
    FloorDivisionOperatorType = "FloorDivision",
    FloorDivisionOperatorMethodType = "FloorDivisionMethod",
    BitwiseAndOperatorType = "BitwiseAnd",
    BitwiseAndOperatorMethodType = "BitwiseAndMethod",
    BitwiseOrOperatorType = "BitwiseOr",
    BitwiseOrOperatorMethodType = "BitwiseOrMethod",
    BitwiseExclusiveOrOperatorType = "BitwiseExclusiveOr",
    BitwiseExclusiveOrOperatorMethodType = "BitwiseExclusiveOrMethod",
    BitwiseLeftShiftOperatorType = "BitwiseLeftShift",
    BitwiseLeftShiftOperatorMethodType = "BitwiseLeftShiftMethod",
    BitwiseRightShiftOperatorType = "BitwiseRightShift",
    BitwiseRightShiftOperatorMethodType = "BitwiseRightShiftMethod",
    ConcatOperatorType = "Concat",
    ConcatOperatorMethodType = "ConcatMethod",
    LessThanOperatorType = "LessThan",
    LessThanOperatorMethodType = "LessThanMethod",
    GreaterThanOperatorType = "GreaterThan",
    GreaterThanOperatorMethodType = "GreaterThanMethod",
    NegationOperatorType = "Negation",
    NegationOperatorMethodType = "NegationMethod",
    BitwiseNotOperatorType = "BitwiseNot",
    BitwiseNotOperatorMethodType = "BitwiseNotMethod",
    LengthOperatorType = "Length",
    LengthOperatorMethodType = "LengthMethod",
    TableNewType = "TableNew",
    TableDeleteType = "TableDelete",
    TableDeleteMethodType = "TableDeleteMethod",
    TableGetType = "TableGet",
    TableGetMethodType = "TableGetMethod",
    TableHasType = "TableHas",
    TableHasMethodType = "TableHasMethod",
    TableSetType = "TableSet",
    TableSetMethodType = "TableSetMethod",
    TableAddKeyType = "TableAddKey",
    TableAddKeyMethodType = "TableAddKeyMethod",
}
const extensionValues: Set<string> = new Set(Object.values(ExtensionKind));

export function getExtensionKindForType(context: TransformationContext, type: ts.Type): ExtensionKind | undefined {
    const value = getPropertyValue(context, type, "__tstlExtension");
    if (value && extensionValues.has(value)) {
        return value as ExtensionKind;
    }
}

const excludedTypeFlags: ts.TypeFlags =
    ((1 << 18) - 1) | // All flags from Any...Never
    ts.TypeFlags.Index |
    ts.TypeFlags.NonPrimitive;

function getPropertyValue(context: TransformationContext, type: ts.Type, propertyName: string): string | undefined {
    if (type.flags & excludedTypeFlags) return;
    const property = type.getProperty(propertyName);
    if (!property) return undefined;
    const propertyType = context.checker.getTypeOfSymbolAtLocation(property, context.sourceFile);
    if (propertyType.isStringLiteral()) return propertyType.value;
}

export function getExtensionKindForNode(context: TransformationContext, node: ts.Node): ExtensionKind | undefined {
    const originalNode = ts.getOriginalNode(node);
    let type = context.checker.getTypeAtLocation(originalNode);
    if (ts.isOptionalChain(originalNode)) {
        type = context.checker.getNonNullableType(type);
    }
    return getExtensionKindForType(context, type);
}

export function getExtensionKindForSymbol(
    context: TransformationContext,
    symbol: ts.Symbol
): ExtensionKind | undefined {
    const type = context.checker.getTypeOfSymbolAtLocation(symbol, context.sourceFile);
    return getExtensionKindForType(context, type);
}

export enum IterableExtensionKind {
    Iterable = "Iterable",
    Pairs = "Pairs",
    PairsKey = "PairsKey",
}

export function getIterableExtensionTypeForType(
    context: TransformationContext,
    type: ts.Type
): IterableExtensionKind | undefined {
    const value = getPropertyValue(context, type, "__tstlIterable");
    if (value && value in IterableExtensionKind) {
        return value as IterableExtensionKind;
    }
}

export function getIterableExtensionKindForNode(
    context: TransformationContext,
    node: ts.Node
): IterableExtensionKind | undefined {
    const type = context.checker.getTypeAtLocation(node);
    return getIterableExtensionTypeForType(context, type);
}
