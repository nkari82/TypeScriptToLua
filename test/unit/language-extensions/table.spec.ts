import * as util from "../../util";
import { invalidCallExtensionUse } from "../../../src/transformation/utils/diagnostics";

describe("LuaTableGet & LuaTableSet extensions", () => {
    test("stand-alone function", () => {
        util.testModule`
            declare const getTable: LuaTableGet<{}, string, number>;
            declare const setTable: LuaTableSet<{}, string, number>;
            const tbl = {};
            setTable(tbl, "foo", 3);
            const result = getTable(tbl, "foo");
            export { result }
        `
            .withLanguageExtensions()
            .setReturnExport("result")
            .expectToEqual(3);
    });

    test("stand-alone function with preceding statements", () => {
        util.testModule`
            declare const setTable: LuaTableSet<{}, string, string>;
            const values: any[] = []
            const tbl: any = {};
            const get = (value: any) => {
                values.push(value)
                return value
            }
            let x = "a"
            setTable(tbl, x += "b", x += "c")
            export const result = tbl.ab
        `
            .withLanguageExtensions()
            .setReturnExport("result")
            .expectToEqual("abc");
    });

    test("namespace function", () => {
        util.testModule`
            declare namespace Table {
                export const get: LuaTableGet<{}, string, number>;
                export const set: LuaTableSet<{}, string, number>;
            }
            const tbl = {};
            Table.set(tbl, "foo", 3);
            const result = Table.get(tbl, "foo");
            export { result }
        `
            .withLanguageExtensions()
            .setReturnExport("result")
            .expectToEqual(3);
    });

    test("method", () => {
        util.testModule`
            interface Table {
                get: LuaTableGetMethod<string, number>;
                set: LuaTableSetMethod<string, number>;
            }
            const tbl = {} as Table;
            tbl.set("foo", 3);
            const result = tbl.get("foo");
            export { result }
        `
            .withLanguageExtensions()
            .setReturnExport("result")
            .expectToEqual(3);
    });

    test.each([
        "const foo: unknown = getTable;",
        "const foo = `${getTable}`;",
        "declare function foo(getTable: LuaTableGet<{}, string, number>): void; foo(getTable);",
        "const foo = (getTable as any)(1, 2);",
        "const foo = [getTable];",
    ])("invalid use (%p)", statement => {
        util.testModule`
            declare const getTable: LuaTableGet<{}, string, number>;
            ${statement}
        `
            .withLanguageExtensions()
            .expectDiagnosticsToMatchSnapshot([invalidCallExtensionUse.code]);
    });
});

describe("LuaTableHas extension", () => {
    test("LuaTableHas standalone function", () => {
        util.testModule`
            declare const tableHas: LuaTableHas<{}, string>;
            const table = { foo: "bar" };

            export const hasFoo = tableHas(table, "foo");
            export const hasBaz = tableHas(table, "baz");
        `
            .withLanguageExtensions()
            .expectToEqual({ hasFoo: true, hasBaz: false });
    });

    test("LuaTableHas nested expression", () => {
        util.testModule`
            declare const tableHas: LuaTableHas<{}, string>;
            const table = { foo: "bar" };

            export const result = \`table has foo: \${tableHas(table, "foo")}\`;
        `
            .withLanguageExtensions()
            .expectToEqual({ result: "table has foo: true" });
    });

    test("LuaTableHas namespace function", () => {
        util.testModule`
            declare namespace Table {
                export const tableHas: LuaTableHas<{}, string>;
            }
            const table = { foo: "bar" };

            export const hasFoo = Table.tableHas(table, "foo");
            export const hasBaz = Table.tableHas(table, "baz");
        `
            .withLanguageExtensions()
            .expectToEqual({ hasFoo: true, hasBaz: false });
    });

    test("LuaTableHasMethod method", () => {
        util.testModule`
            interface TableWithHas {
                has: LuaTableHasMethod<string>;
                set: LuaTableSetMethod<string, number>;
            }
            const table = {} as TableWithHas;
            table.set("foo", 42);

            export const hasFoo = table.has("foo");
            export const hasBaz = table.has("baz");
        `
            .withLanguageExtensions()
            .expectToEqual({ hasFoo: true, hasBaz: false });
    });

    test("LuaTableHas as statement", () => {
        util.testModule`
            declare const tableHas: LuaTableHas<{}, string>;
            const table = { foo: "bar" };

            let count = 0;
            function getKey() {
                count++;
                return "foo";
            }

            // The result is not captured, but the expression will be evaulated
            tableHas(table, getKey());

            export const numCalls = count;
        `
            .withLanguageExtensions()
            .expectToEqual({ numCalls: 1 });
    });

    test.each([
        "const foo: unknown = tableHas;",
        "const foo = `${tableHas}`;",
        "declare function foo(tableHas: LuaTableHas<{}, string>): void; foo(tableHas);",
        "const foo = (tableHas as any)(1, 2);",
        "const foo = [tableHas];",
    ])("invalid use (%p)", statement => {
        util.testModule`
            declare const tableHas: LuaTableHas<{}, string>;
            ${statement}
        `
            .withLanguageExtensions()
            .expectDiagnosticsToMatchSnapshot([invalidCallExtensionUse.code]);
    });

    test.each(["LuaTable<string, number>", "LuaMap<string, number>", "LuaSet<string>"])(
        "invalid use method assignment (%p)",
        type => {
            util.testModule`
                const table = new ${type}();
                const has = table.has;
            `
                .withLanguageExtensions()
                .expectDiagnosticsToMatchSnapshot([invalidCallExtensionUse.code]);
        }
    );

    test.each(["LuaTable<string, number>", "LuaMap<string, number>", "LuaSet<string>"])(
        "invalid use method expression (%p)",
        type => {
            util.testModule`
                const table = new ${type}();
                ["a", "b", "c"].map(table.has);
            `
                .withLanguageExtensions()
                .expectDiagnosticsToMatchSnapshot([invalidCallExtensionUse.code]);
        }
    );
});

describe("LuaTableDelete extension", () => {
    test("LuaTableDelete standalone function", () => {
        util.testModule`
            declare const tableDelete: LuaTableDelete<{}, string>;

            export const table = { foo: "bar", baz: "baz" };
            tableDelete(table, "foo");
        `
            .withLanguageExtensions()
            .expectToEqual({ table: { baz: "baz" } });
    });

    test("LuaTableDelete namespace function", () => {
        util.testModule`
            declare namespace Table {
                export const tableDelete: LuaTableDelete<{}, string>;
            }

            export const table = { foo: "bar", baz: "baz" };
            Table.tableDelete(table, "foo");
        `
            .withLanguageExtensions()
            .expectToEqual({ table: { baz: "baz" } });
    });

    test("LuaTableHasMethod method", () => {
        util.testModule`
            interface TableWithDelete {
                delete: LuaTableDeleteMethod<string>;
                set: LuaTableSetMethod<string, number>;
            }
            export const table = {} as TableWithDelete;
            table.set("foo", 42);
            table.set("bar", 12);
            table.delete("foo");
        `
            .withLanguageExtensions()
            .expectToEqual({ table: { bar: 12 } });
    });

    test.each([
        ["LuaTableDelete<{}, string>", 'func({}, "foo")', true],
        ["LuaTableDelete<{}, string>", '"truthy" && func({}, "foo")', true],
        ["LuaTableSet<{}, string, number>", 'func({}, "foo", 3)', undefined],
    ])("Table functions used as expression", (funcType, expression, value) => {
        util.testModule`
            declare const func: ${funcType}
            export const result = ${expression}
        `
            .withLanguageExtensions()
            .setReturnExport("result")
            .expectToEqual(value);
    });

    test.each([
        ["LuaTableDeleteMethod<string>", 'tbl.func("foo")', true],
        ["LuaTableSetMethod<string, number>", 'tbl.func("foo", 3)', undefined],
    ])("Table methods used as expression", (funcType, expression, value) => {
        util.testModule`
            const tbl = {} as { func: ${funcType} }
            export const result = ${expression}
        `
            .withLanguageExtensions()
            .setReturnExport("result")
            .expectToEqual(value);
    });
});

describe("LuaTableAddKey extension", () => {
    test("LuaTableAddKey standalone function", () => {
        util.testModule`
            declare const tableAddKey: LuaTableAddKey<{}, string>;
            export const table = { foo: "bar" };
            tableAddKey(table, "baz");
        `
            .withLanguageExtensions()
            .expectToEqual({ table: { foo: "bar", baz: true } });
    });

    test("LuaTableAddKey namespace function", () => {
        util.testModule`
            declare namespace Table {
                export const tableAddKey: LuaTableAddKey<{}, string>;
            }
            export const table = { foo: "bar" };
            Table.tableAddKey(table, "baz");
        `
            .withLanguageExtensions()
            .expectToEqual({ table: { foo: "bar", baz: true } });
    });

    test("LuaTableAddKey method", () => {
        util.testModule`
            interface TableWithAddKey {
                addKey: LuaTableAddKeyMethod<string>;
            }
            export const table = {} as TableWithAddKey;
            table.addKey("bar");
        `
            .withLanguageExtensions()
            .expectToEqual({ table: { bar: true } });
    });

    test.each([
        ["LuaTableAddKey<{}, string>", 'func({}, "foo")', undefined],
        ["LuaTableAddKey<{}, string>", '"truthy" && func({}, "foo")', undefined],
    ])("Table functions used as expression", (funcType, expression, value) => {
        util.testModule`
            declare const func: ${funcType}
            export const result = ${expression}
        `
            .withLanguageExtensions()
            .setReturnExport("result")
            .expectToEqual(value);
    });

    test("Method used as expression", () => {
        util.testModule`
            const tbl = {} as { func: LuaTableAddKeyMethod<string> }
            export const result = tbl.func("foo")
        `
            .withLanguageExtensions()
            .setReturnExport("result")
            .expectToEqual(undefined);
    });
});

describe("LuaTable extension interface", () => {
    test("untyped table", () => {
        util.testFunction`
            const tbl = new LuaTable();
            tbl.set("foo", 3);
            return tbl.get("foo");
        `
            .withLanguageExtensions()
            .expectToEqual(3);
    });

    test("typed table", () => {
        util.testFunction`
            const tbl = new LuaTable<string, number>();
            tbl.set("foo", 3);
            return tbl.get("foo");
        `
            .withLanguageExtensions()
            .expectToEqual(3);
    });

    // Test to catch issue from https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1033
    test("typed table in type declaration", () => {
        util.testFunction`
            function fill(tbl: LuaTable<string, number>) {
                tbl.set("foo", 3);
                return tbl;
            }
            return fill(new LuaTable()).get("foo");
        `
            .withLanguageExtensions()
            .expectToEqual(3);
    });

    test.each([["null"], ["undefined"], ["number | undefined"], ["string | null"], ["unknown"]])(
        "LuaTable in strict mode does not accept key type that could be nil (%p)",
        keyType => {
            util.testExpression`new LuaTable<${keyType}, unknown>()`
                .withLanguageExtensions()
                .setOptions({ strict: true })
                .expectToHaveDiagnostics()
                .expectDiagnosticsToMatchSnapshot();
        }
    );

    test("object keyed table", () => {
        util.testFunction`
            interface Key { keyStr: string }
            const key: Key = {keyStr: "foo"};
            const tbl = new LuaTable<Key, number>();
            tbl.set(key, 3);
            return tbl.get(key);
        `
            .withLanguageExtensions()
            .expectToEqual(3);
    });

    test("table length", () => {
        util.testFunction`
            const tbl = new LuaTable<number, string>();
            tbl.set(1, "foo");
            tbl.set(3, "bar");
            return tbl.length();
        `
            .withLanguageExtensions()
            .expectToEqual(1);
    });

    test("table has", () => {
        util.testFunction`
            const tbl = new LuaTable<number, string>();
            tbl.set(3, "foo");
            return [tbl.has(1), tbl.has(3)];
        `
            .withLanguageExtensions()
            .expectToEqual([false, true]);
    });

    test("table delete", () => {
        util.testFunction`
            const tbl = new LuaTable<string, number>();
            tbl.set("foo", 1);
            tbl.set("bar", 3);
            tbl.set("baz", 5);
            tbl.delete("bar");
            tbl.delete("foo");
            return tbl;
        `
            .withLanguageExtensions()
            .expectToEqual({ baz: 5 });
    });

    test("table add", () => {
        util.testFunction`
            const tbl = new LuaSet<string>()
            tbl.add("foo");
            return tbl
        `
            .withLanguageExtensions()
            .expectToEqual({ foo: true });
    });

    test.each(['new LuaTable().get("foo");', 'new LuaTable().set("foo", "bar");'])(
        "table immediate access (%p)",
        statement => {
            util.testFunction(statement).withLanguageExtensions().expectToHaveNoDiagnostics();
        }
    );

    test("table pairs iterate", () => {
        util.testFunction`
            const tbl = new LuaTable<string, number>();
            tbl.set("foo", 1);
            tbl.set("bar", 3);
            tbl.set("baz", 5);
            const results: Record<string, number> = {};
            for (const [k, v] of tbl) {
                results[k] = v;
            }
            return results;
        `
            .withLanguageExtensions()
            .expectToEqual({ foo: 1, bar: 3, baz: 5 });
    });
});

test.each([
    [undefined, undefined],
    ["new LuaSet()", true],
])("call on optional table with strictNullChecks (%s)", (value, expected) => {
    util.testFunction`
        function getFoo(): LuaSet<string> | undefined {
            return ${value}
        }
        const foo = getFoo()
        foo?.add("foo")
        return foo?.has("foo")
    `
        .setOptions({
            strictNullChecks: true,
        })
        .withLanguageExtensions()
        .expectToEqual(expected);
});
