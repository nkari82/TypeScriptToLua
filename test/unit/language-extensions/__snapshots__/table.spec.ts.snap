// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("null"): code 1`] = `
"local ____exports = {}
____exports.__result = {}
return ____exports"
`;

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("null"): diagnostics 1`] = `"main.ts(1,38): error TS2344: Type 'null' does not satisfy the constraint 'AnyNotNil'."`;

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("number | undefined"): code 1`] = `
"local ____exports = {}
____exports.__result = {}
return ____exports"
`;

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("number | undefined"): diagnostics 1`] = `
"main.ts(1,38): error TS2344: Type 'number | undefined' does not satisfy the constraint 'AnyNotNil'.
  Type 'undefined' is not assignable to type 'AnyNotNil'."
`;

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("string | null"): code 1`] = `
"local ____exports = {}
____exports.__result = {}
return ____exports"
`;

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("string | null"): diagnostics 1`] = `
"main.ts(1,38): error TS2344: Type 'string | null' does not satisfy the constraint 'AnyNotNil'.
  Type 'null' is not assignable to type 'AnyNotNil'."
`;

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("undefined"): code 1`] = `
"local ____exports = {}
____exports.__result = {}
return ____exports"
`;

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("undefined"): diagnostics 1`] = `"main.ts(1,38): error TS2344: Type 'undefined' does not satisfy the constraint 'AnyNotNil'."`;

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("unknown"): code 1`] = `
"local ____exports = {}
____exports.__result = {}
return ____exports"
`;

exports[`LuaTable extension interface LuaTable in strict mode does not accept key type that could be nil ("unknown"): diagnostics 1`] = `"main.ts(1,38): error TS2344: Type 'unknown' does not satisfy the constraint 'AnyNotNil'."`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("const foo = (getTable as any)(1, 2);"): code 1`] = `"foo = getTable(_G, 1, 2)"`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("const foo = (getTable as any)(1, 2);"): diagnostics 1`] = `"main.ts(3,26): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("const foo = [getTable];"): code 1`] = `"foo = {getTable}"`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("const foo = [getTable];"): diagnostics 1`] = `"main.ts(3,26): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("const foo = \`\${getTable}\`;"): code 1`] = `"foo = tostring(getTable)"`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("const foo = \`\${getTable}\`;"): diagnostics 1`] = `"main.ts(3,28): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("const foo: unknown = getTable;"): code 1`] = `"foo = getTable"`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("const foo: unknown = getTable;"): diagnostics 1`] = `"main.ts(3,34): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("declare function foo(getTable: LuaTableGet<{}, string, number>): void; foo(getTable);"): code 1`] = `"foo(_G, getTable)"`;

exports[`LuaTableGet & LuaTableSet extensions invalid use ("declare function foo(getTable: LuaTableGet<{}, string, number>): void; foo(getTable);"): diagnostics 1`] = `"main.ts(3,88): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use ("const foo = (tableHas as any)(1, 2);"): code 1`] = `"foo = tableHas(_G, 1, 2)"`;

exports[`LuaTableHas extension invalid use ("const foo = (tableHas as any)(1, 2);"): diagnostics 1`] = `"main.ts(3,26): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use ("const foo = [tableHas];"): code 1`] = `"foo = {tableHas}"`;

exports[`LuaTableHas extension invalid use ("const foo = [tableHas];"): diagnostics 1`] = `"main.ts(3,26): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use ("const foo = \`\${tableHas}\`;"): code 1`] = `"foo = tostring(tableHas)"`;

exports[`LuaTableHas extension invalid use ("const foo = \`\${tableHas}\`;"): diagnostics 1`] = `"main.ts(3,28): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use ("const foo: unknown = tableHas;"): code 1`] = `"foo = tableHas"`;

exports[`LuaTableHas extension invalid use ("const foo: unknown = tableHas;"): diagnostics 1`] = `"main.ts(3,34): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use ("declare function foo(tableHas: LuaTableHas<{}, string>): void; foo(tableHas);"): code 1`] = `"foo(_G, tableHas)"`;

exports[`LuaTableHas extension invalid use ("declare function foo(tableHas: LuaTableHas<{}, string>): void; foo(tableHas);"): diagnostics 1`] = `"main.ts(3,80): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use method assignment ("LuaMap<string, number>"): code 1`] = `
"____table = {}
has = ____table.has"
`;

exports[`LuaTableHas extension invalid use method assignment ("LuaMap<string, number>"): diagnostics 1`] = `"main.ts(3,29): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use method assignment ("LuaSet<string>"): code 1`] = `
"____table = {}
has = ____table.has"
`;

exports[`LuaTableHas extension invalid use method assignment ("LuaSet<string>"): diagnostics 1`] = `"main.ts(3,29): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use method assignment ("LuaTable<string, number>"): code 1`] = `
"____table = {}
has = ____table.has"
`;

exports[`LuaTableHas extension invalid use method assignment ("LuaTable<string, number>"): diagnostics 1`] = `"main.ts(3,29): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use method expression ("LuaMap<string, number>"): code 1`] = `
"local ____lualib = require(\\"lualib_bundle\\")
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
____table = {}
__TS__ArrayMap({\\"a\\", \\"b\\", \\"c\\"}, ____table.has)"
`;

exports[`LuaTableHas extension invalid use method expression ("LuaMap<string, number>"): diagnostics 1`] = `"main.ts(3,37): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use method expression ("LuaSet<string>"): code 1`] = `
"local ____lualib = require(\\"lualib_bundle\\")
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
____table = {}
__TS__ArrayMap({\\"a\\", \\"b\\", \\"c\\"}, ____table.has)"
`;

exports[`LuaTableHas extension invalid use method expression ("LuaSet<string>"): diagnostics 1`] = `"main.ts(3,37): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`LuaTableHas extension invalid use method expression ("LuaTable<string, number>"): code 1`] = `
"local ____lualib = require(\\"lualib_bundle\\")
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
____table = {}
__TS__ArrayMap({\\"a\\", \\"b\\", \\"c\\"}, ____table.has)"
`;

exports[`LuaTableHas extension invalid use method expression ("LuaTable<string, number>"): diagnostics 1`] = `"main.ts(3,37): error TSTL: This function must be called directly and cannot be referred to."`;
